### PR TITLE
IdenticalEqual and NotIdenticalNotEqual mutators

### PR DESCRIPTION
**Pull request by sanmai**
This PR:

- [x] Adds IdenticalEqual mutator (replaces "===" with "==")
- [x] Adds NotIdenticalNotEqual mutator (replaces "!==" with "!=")
- [x] Covered by tests
- [x] E2E fixtures updated
- [x] Doc PR: https://github.com/infection/site/pull/35
- [x] Code updated to match minimal MSI  (should be 86%)

### CODE REVIEW

**Comment by borNfreee** [comment]
> I think it should be possible to get MSI back up to 86%

That would be great. In https://github.com/infection/infection/pull/297 (still WIP), I tried to increase covered code MSI for each mutator to 100%.

It is useful because

* trying to kill the Mutant gives you better understanding about usefulness of a particular mutator
* it forces you to write more tests and remove or update invalid/redundant code

**Comment by sanmai** [comment]
Where we have to use the strict `===` identical comparison:
* If we want to distinguish between `null`, `false`, `0`, and between `1` and `true`; `strpos` is a very well known companion of the identity operator.
* If we can expect type juggling: string vs number comparison, float vs. int, exact class instances.

Where we can use the strict `===` identical comparison, but it is useless and needless:
* Where no type juggling is expected, string vs string comparison being one of common offenders.
* Where input types are known beforehand, e.g. `function (float $v) {return $v == 1;}`

File: src/Config/Guesser/PhpUnitPathGuesser.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ private function getPhpUnitDir(array $parsedPaths)
     {
         foreach ($parsedPaths as $namespace => $parsedPath) {
             // for old Symfony prjects (<=2.7) phpunit.xml is located in ./app folder
-            if (strpos($namespace, 'SymfonyStandard') !== false && trim($parsedPath, '/') === 'app') {
+            if (strpos($namespace, 'SymfonyStandard') !== false && trim($parsedPath, '/') == 'app') {
```
Comment (sanmai) [inline src/Config/Guesser/PhpUnitPathGuesser.php]: Redundant string vs string comparison.

File: src/Finder/Iterator/RealPathFilterIterator.php [hunk 2]
```diff
@@ -22,7 +22,7 @@ public function accept()
     {
         $filename = $this->current()->getRealPath();
 
-        if ('\\' === DIRECTORY_SEPARATOR) {
+        if ('\\' == DIRECTORY_SEPARATOR) {
```
Comment (sanmai) [inline src/Finder/Iterator/RealPathFilterIterator.php]: string vs string - redundant

File: src/Finder/SourceFilesFinder.php [hunk 3]
```diff
@@ -45,7 +45,7 @@ public function getSourceFiles(string $filter = ''): Finder
 
         $this->in($this->sourceDirectories)->files();
 
-        if ('' === $filter) {
+        if ('' == $filter) {
```
Comment (sanmai) [inline src/Finder/SourceFilesFinder.php]: string vs string - redundant

File: src/Config/ValueProvider/ExcludeDirsProvider.php [hunk 4]
```diff
@@ -62,15 +62,15 @@ public function get(InputInterface $input, OutputInterface $output, array $dirsI
         );
         $excludedDirs = [];
 
-        if ($sourceDirs === ['.']) {
+        if ($sourceDirs == ['.']) {
```
Comment (sanmai) [inline src/Config/ValueProvider/ExcludeDirsProvider.php]: $sourceDirs always an array here

File: src/Config/ValueProvider/ExcludeDirsProvider.php [hunk 5]
```diff
@@ -62,15 +62,15 @@ public function get(InputInterface $input, OutputInterface $output, array $dirsI
...
             foreach (self::EXCLUDED_ROOT_DIRS as $dir) {
                 if (in_array($dir, $dirsInCurrentDir, true)) {
                     $excludedDirs[] = $dir;
                 }
             }
 
             $autocompleteValues = $dirsInCurrentDir;
-        } elseif (count($sourceDirs) === 1) {
+        } elseif (count($sourceDirs) == 1) {
```
Comment (sanmai) [inline src/Config/ValueProvider/ExcludeDirsProvider.php]: count() returns only ints

File: src/Config/ValueProvider/TestFrameworkConfigPathProvider.php [hunk 6]
```diff
@@ -49,7 +49,7 @@ public function get(InputInterface $input, OutputInterface $output, array $dirsI
 
             return null;
         } catch (\Exception $e) {
-            if ($testFramework !== TestFrameworkTypes::PHPUNIT) {
+            if ($testFramework != TestFrameworkTypes::PHPUNIT) {
```
Comment (sanmai) [inline src/Config/ValueProvider/TestFrameworkConfigPathProvider.php]: Redundant string vs string

File: src/Mutator/Boolean/FalseValue.php [hunk 7]
```diff
@@ -32,6 +32,6 @@ protected function mutatesNode(Node $node): bool
             return false;
         }
 
-        return strtolower($node->name->getFirst()) === 'false';
+        return strtolower($node->name->getFirst()) == 'false';
```
Comment (sanmai) [inline src/Mutator/Boolean/FalseValue.php]: Redundant string vs string

File: src/Mutator/Boolean/TrueValue.php [hunk 8]
```diff
@@ -32,6 +32,6 @@ protected function mutatesNode(Node $node): bool
             return false;
         }
 
-        return strtolower($node->name->getFirst()) === 'true';
+        return strtolower($node->name->getFirst()) == 'true';
```
Comment (sanmai) [inline src/Mutator/Boolean/TrueValue.php]: Redundant string vs string

File: src/Mutator/Number/DecrementInteger.php [hunk 9]
```diff
@@ -28,6 +28,6 @@ public function mutate(Node $node)
 
     protected function mutatesNode(Node $node): bool
     {
-        return $node instanceof Node\Scalar\LNumber && $node->value !== 1;
+        return $node instanceof Node\Scalar\LNumber && $node->value != 1;
```
Comment (sanmai) [inline src/Mutator/Number/DecrementInteger.php]: We know the type

File: src/Mutator/Number/IncrementInteger.php [hunk 10]
```diff
@@ -28,6 +28,6 @@ public function mutate(Node $node)
 
     protected function mutatesNode(Node $node): bool
     {
-        return $node instanceof Node\Scalar\LNumber && $node->value !== 0;
+        return $node instanceof Node\Scalar\LNumber && $node->value != 0;
```
Comment (sanmai) [inline src/Mutator/Number/IncrementInteger.php]: Type is known

File: src/Mutator/Number/OneZeroFloat.php [hunk 11]
```diff
@@ -32,6 +32,6 @@ public function mutate(Node $node)
 
     protected function mutatesNode(Node $node): bool
     {
-        return $node instanceof Node\Scalar\DNumber && ($node->value === 0.0 || $node->value === 1.0);
+        return $node instanceof Node\Scalar\DNumber && ($node->value == 0.0 || $node->value == 1.0);
```
Comment (sanmai) [inline src/Mutator/Number/OneZeroFloat.php]: Always float, never int here

File: src/Mutator/Number/OneZeroInteger.php [hunk 12]
```diff
@@ -23,7 +23,7 @@
      */
     public function mutate(Node $node)
     {
-        if ($node->value === 0) {
+        if ($node->value == 0) {
```
Comment (sanmai) [inline src/Mutator/Number/OneZeroInteger.php]: Always an int here

File: src/Mutator/ReturnValue/This.php [hunk 13]
```diff
@@ -32,7 +32,7 @@ protected function mutatesNode(Node $node): bool
     {
         return $node instanceof Node\Stmt\Return_ &&
             $node->expr instanceof Node\Expr\Variable &&
-            $node->expr->name === 'this'
+            $node->expr->name == 'this'
```
Comment (sanmai) [inline src/Mutator/ReturnValue/This.php]: Redundant string vs string

File: src/Mutator/Util/MutatorsGenerator.php [hunk 14]
```diff
@@ -98,7 +98,7 @@ private function registerFromClass(string $mutator, $setting)
             if (!array_key_exists($mutator, $this->mutatorList)) {
                 $this->mutatorList[$mutator] = [];
             }
-        } elseif (!array_key_exists($mutator, $this->mutatorList) || $this->mutatorList[$mutator] === []) {
+        } elseif (!array_key_exists($mutator, $this->mutatorList) || !$this->mutatorList[$mutator]) {
```
Comment (sanmai) [inline src/Mutator/Util/MutatorsGenerator.php]: `[]` is same as `false`, and `mutatorList` is always a 2D array

File: src/Process/Builder/ProcessBuilder.php [hunk 15]
```diff
@@ -49,7 +49,7 @@ public function getProcessForInitialTestRun(
         bool $skipCoverage,
         array $phpExtraOptions = []
     ): Process {
-        $includeArgs = PHP_SAPI === 'phpdbg' || $skipCoverage;
+        $includeArgs = PHP_SAPI == 'phpdbg' || $skipCoverage;
```
Comment (sanmai) [inline src/Process/Builder/ProcessBuilder.php]: Redundant string vs string

File: src/Process/Listener/MutationTestingResultsLoggerSubscriber.php [hunk 16]
```diff
@@ -92,7 +92,7 @@ public function onMutationTestingFinished(MutationTestingFinished $event)
     private function filterLogTypes(array $logTypes): array
     {
         foreach ($logTypes as $key => $value) {
-            if ($this->logVerbosity === LogVerbosity::NONE) {
+            if ($this->logVerbosity == LogVerbosity::NONE) {
```
Comment (sanmai) [inline src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: No ambiguity here

File: src/TestFramework/Coverage/CodeCoverageData.php [hunk 17]
```diff
@@ -147,7 +147,7 @@ public function getAllTestsFor(MutationInterface $mutation): array
      */
     private function getCoverage(): array
     {
-        if (null === $this->coverage) {
+        if (!isset($this->coverage)) {
```
Comment (sanmai) [inline src/TestFramework/Coverage/CodeCoverageData.php]: Unavoidable

File: src/TestFramework/Coverage/CodeCoverageData.php [hunk 18]
```diff
@@ -171,7 +171,7 @@ private function getCoverage(): array
 
     private function addTestExecutionInfo(array $coverage): array
     {
-        if ($this->testFileDataProvider === null) {
+        if (!$this->testFileDataProvider) {
```
Comment (sanmai) [inline src/TestFramework/Coverage/CodeCoverageData.php]: `null` or an object, but any object is always `true`

File: src/TestFramework/Factory.php [hunk 19]
```diff
@@ -81,7 +81,7 @@ public function __construct(
 
     public function create(string $adapterName, bool $skipCoverage): AbstractTestFrameworkAdapter
     {
-        if ($adapterName === TestFrameworkTypes::PHPUNIT) {
+        if ($adapterName == TestFrameworkTypes::PHPUNIT) {
```
Comment (sanmai) [inline src/TestFramework/Factory.php]: string v string

File: src/TestFramework/PhpUnit/Config/Builder/InitialConfigBuilder.php [hunk 20]
```diff
@@ -121,7 +121,7 @@ private function addCoverageFilterWhitelistIfDoesNotExist(\DOMDocument $dom, \DO
     {
         $filterNode = $this->getNode($dom, $xPath, 'filter');
 
-        if ($filterNode === null) {
+        if (!$filterNode) {
```
Comment (sanmai) [inline src/TestFramework/PhpUnit/Config/Builder/InitialConfigBuilder.php]: any object is always `true`

File: src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php [hunk 21]
```diff
@@ -75,14 +75,14 @@ private function processXmlFileCoverage(string $relativeCoverageFilePath, string
...
-        if ($percentage === '0.00' || empty($percentage)) {
+        if (!$percentage) {
             return [$sourceFilePath => $defaultCoverageFileData];
         }
 
         /** @var \DOMNodeList $lineCoverageNodes */
         $lineCoverageNodes = $xPath->query('/phpunit/file/coverage/line');
 
-        if ($lineCoverageNodes->length === 0) {
+        if (!$lineCoverageNodes->length) {
```
Comment (sanmai) [inline src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: Never `false`

File: src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php [hunk 22]
```diff
@@ -156,7 +156,7 @@ private function getCoveredLinesData(\DOMNodeList $lineCoverageNodes): array
             $lineNumber = (int) $lineCoverageNode->getAttribute('nr');
 
             foreach ($lineCoverageNode->childNodes as $coveredNode) {
-                if ($coveredNode->nodeName === 'covered') {
+                if ($coveredNode->nodeName == 'covered') {
```
Comment (sanmai) [inline src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: string v string

File: src/TestFramework/PhpUnit/Coverage/PhpUnitTestFileDataProvider.php [hunk 23]
```diff
@@ -36,7 +36,7 @@ public function getTestFileInfo(string $fullyQualifiedClassName): array
 
         $nodes = $xPath->query(sprintf('//testsuite[@name="%s"]', $fullyQualifiedClassName));
 
-        if ($nodes->length === 0) {
+        if (!$nodes->length) {
```
Comment (sanmai) [inline src/TestFramework/PhpUnit/Coverage/PhpUnitTestFileDataProvider.php]: Never `false`

File: src/TestFramework/PhpUnit/Coverage/PhpUnitTestFileDataProvider.php [hunk 24]
```diff
@@ -48,7 +48,7 @@ public function getTestFileInfo(string $fullyQualifiedClassName): array
 
     private function getXPath(): \DOMXPath
     {
-        if ($this->xPath === null) {
+        if (!$this->xPath) {
```
Comment (sanmai) [inline src/TestFramework/PhpUnit/Coverage/PhpUnitTestFileDataProvider.php]: any object is `true`

File: src/Visitor/FullyQualifiedClassNameVisitor.php [hunk 25]
```diff
@@ -28,9 +28,7 @@ public function enterNode(Node $node)
         if ($node instanceof Stmt\Namespace_) {
             $this->namespace = $node->name;
         } elseif ($node instanceof Stmt\ClassLike) {
-            $node->fullyQualifiedClassName = null === $node->name
-                ? null
-                : Name::concat($this->namespace, $node->name->name);
+            $node->fullyQualifiedClassName = $node->name ? Name::concat($this->namespace, $node->name->name) : null;
```
Comment (sanmai) [inline src/Visitor/FullyQualifiedClassNameVisitor.php]: node name is always a string that never starts with a digit

File: src/Visitor/MutatorVisitor.php [hunk 26]
```diff
@@ -37,10 +37,10 @@ public function leaveNode(Node $node)
         $mutator = $this->mutation->getMutator();
         $mutatedAttributes = $this->mutation->getAttributes();
 
-        $isEqualPosition = $attributes['startTokenPos'] === $mutatedAttributes['startTokenPos'] &&
-            $attributes['endTokenPos'] === $mutatedAttributes['endTokenPos'];
+        $isEqualPosition = $attributes['startTokenPos'] == $mutatedAttributes['startTokenPos'] &&
+            $attributes['endTokenPos'] == $mutatedAttributes['endTokenPos'];
```
Comment (sanmai) [inline src/Visitor/MutatorVisitor.php]: Same type comparison

File: src/Visitor/MutatorVisitor.php [hunk 27]
```diff
@@ -37,10 +37,10 @@ public function leaveNode(Node $node)
...
 
-        if ($isEqualPosition && $this->mutation->getMutatedNodeClass() === get_class($node)) {
+        if ($isEqualPosition && $this->mutation->getMutatedNodeClass() == get_class($node)) {
```
Comment (sanmai) [inline src/Visitor/MutatorVisitor.php]: string v string

File: src/Visitor/ReflectionVisitor.php [hunk 28]
```diff
@@ -41,7 +41,7 @@ public function beforeTraverse(array $nodes)
 
     public function enterNode(Node $node)
     {
-        if ($node instanceof Node\Stmt\ClassLike && $node->fullyQualifiedClassName !== null) {
+        if ($node instanceof Node\Stmt\ClassLike && isset($node->fullyQualifiedClassName)) {
```
Comment (sanmai) [inline src/Visitor/ReflectionVisitor.php]: This is an injected property, original object does not necessarily have it

File: src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php [hunk 29]
```diff
@@ -75,14 +75,14 @@ private function processXmlFileCoverage(string $relativeCoverageFilePath, string
 
         $defaultCoverageFileData = ['byLine' => [], 'byMethod' => []];
 
-        if ($percentage === '0.00' || empty($percentage)) {
+        if (!$percentage) {
```
Comment (borNfreee) [inline src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: seems like this is not identical

`'0.00'` string is not falsy, but truthy https://3v4l.org/osjdM

Comment (sanmai) [reply src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: Oh well. I'll see if I can add a test for this variant.

Comment (sanmai) [reply src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: No easy way to test this unfortunately. I seems like 261a483b is the best course of action here. WDYT?

**Comment by borNfreee** [comment]
Ok, let's get it merged. Thank you @sanmai

**Pull request merged**